File: matlab/tests/testPsrvel.m
function testPsrvel()
%TESTPSRVEL Regression checks for MATLAB PSRVEL support.

sample = ['#PSRVELA,USB1,0,51.5,FINESTEERING,2209,511827.000,' ...
    '02000020,0dd6,16809;SOL_COMPUTED,WAAS,0.000,4.000,0.0175,' ...
    '290.743174,0.0309,0*3d24adcc'];
r = gnssparser.novatel.parsePsrvelLine(sample, true);
assert(r.week == 2209);
assert(abs(r.sow - 511827.0) < 1e-12);
assert(strcmp(r.sol_status, 'SOL_COMPUTED'));
assert(strcmp(r.vel_type, 'WAAS'));
assert(abs(r.hor_speed_mps - 0.0175) < 1e-12);
assert(abs(r.vert_speed_mps - 0.0309) < 1e-12);

noncomputed = ['#PSRVELA,COM1,0,0.0,FINESTEERING,2209,1.000,' ...
    '00000000,0000,16809;INSUFFICIENT_OBS,NONE,0,0,0,0,0,0*00000000'];
r2 = gnssparser.novatel.parsePsrvelLine(noncomputed, false);
assert(strcmp(r2.sol_status, 'INSUFFICIENT_OBS'));

file = [tempname '.log'];
fid = fopen(file, 'w');
assert(fid >= 0);
cleanup = onCleanup(@() cleanupFile(fid, file)); %#ok<NASGU>
fprintf(fid, 'noise\n%s\n', sample);
fclose(fid);
fid = -1;
[records, stats] = readNovatelPsrvel(file, 'VerifyCrc', true);
assert(numel(records) == 1);
assert(stats.records == 1 && stats.target_lines == 1);

fprintf('testPsrvel: PASS\n');
end

function cleanupFile(fid, file)
if fid >= 0, fclose(fid); end
if exist(file, 'file'), delete(file); end
end

File: tests/test_psrvel.py
import io
import sys
import unittest
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parents[1] / "python"))

from gnss_parser.novatel import (
    NovatelAsciiParseError,
    iter_psrvel,
    parse_psrvel_line,
    read_psrvel,
)

SAMPLE = "#PSRVELA,USB1,0,51.5,FINESTEERING,2209,511827.000,02000020,0dd6,16809;SOL_COMPUTED,WAAS,0.000,4.000,0.0175,290.743174,0.0309,0*3d24adcc"


class TestPsrvel(unittest.TestCase):
    def test_parse_official_sample(self):
        record = parse_psrvel_line(SAMPLE, verify_crc=True)
        self.assertEqual(record.sol_status, "SOL_COMPUTED")
        self.assertEqual(record.vel_type, "WAAS")
        self.assertEqual(record.latency_s, 0.0)
        self.assertEqual(record.age_s, 4.0)
        self.assertAlmostEqual(record.hor_speed_mps, 0.0175)
        self.assertAlmostEqual(record.track_deg, 290.743174)
        self.assertAlmostEqual(record.vert_speed_mps, 0.0309)
        self.assertEqual(record.week, 2209)
        self.assertEqual(record.sow, 511827.0)
        self.assertEqual(record.time_status, "FINESTEERING")

    def test_mixed_log_exact_matching(self):
        text = "\n".join(
            [
                "$GPRMC,unrelated",
                SAMPLE.replace("#PSRVELA", "#PSRVEL2A"),
                SAMPLE,
            ]
        )
        records = read_psrvel(io.StringIO(text), verify_crc=True)
        self.assertEqual(len(records), 1)

    def test_tolerant_skips_malformed_target(self):
        bad = SAMPLE.replace(",0.0309,0*", ",BAD,0*")
        records = list(iter_psrvel(io.StringIO(bad + "\n" + SAMPLE)))
        self.assertEqual(len(records), 1)

    def test_strict_reports_line_number(self):
        bad = SAMPLE.replace(",0.0309,0*", ",BAD,0*")
        with self.assertRaisesRegex(NovatelAsciiParseError, "line 2:"):
            list(iter_psrvel(io.StringIO("$GPRMC,x\n" + bad), strict=True))

    def test_no_hidden_solution_filter(self):
        bad_status = SAMPLE.replace(
            "SOL_COMPUTED,WAAS", "INSUFFICIENT_OBS,NONE"
        )
        record = parse_psrvel_line(bad_status)
        self.assertEqual(record.sol_status, "INSUFFICIENT_OBS")
        self.assertEqual(record.vel_type, "NONE")

    def test_wrong_body_field_count(self):
        truncated = SAMPLE.replace(",0.0309,0*", ",0.0309*")
        with self.assertRaisesRegex(
            NovatelAsciiParseError, "requires 8 body fields"
        ):
            parse_psrvel_line(truncated)


if __name__ == "__main__":
    unittest.main()
